File: 2019/elixir/1/main.exs
defmodule D1 do
  def one() do
    getInput("input.txt")
    |> Enum.map(&calcFuel/1)
    |> Enum.reduce(0, fn x, acc -> x + acc end)
  end

  def two() do
    getInput("input.txt")
    |> Enum.map(fn x -> 
      x
      |> calcFuel
      |> calcFuelOfFuel(0) 
      end)
    |> Enum.reduce(0, fn x, acc -> x + acc end)
  end

  def getInput(filename) do
    {:ok, input} = File.read(filename)
    input
    |> String.split("\n")
    |> Enum.map(&String.to_integer/1)
  end

  def calcFuel(mass) do
    mass
    |> Math.divide(3)
    |> Float.floor
    |> Math.substract(2)
    |> trunc
  end

  def calcFuelOfFuel(mass, acc) when mass > 0 do
    acc = mass
    |> calcFuel
    |> calcFuelOfFuel(acc)
    |> Math.addition(mass)
    acc
  end

  def calcFuelOfFuel(_, acc) do
    acc
  end
end

defmodule Math do
  def divide(a, b) do
    a / b
  end

  def substract(a, b) do
    a - b
  end

  def addition(a, b) do
    a + b
  end
end

IO.puts("One: #{D1.one()}")
# D1.two()
IO.puts("Two: #{D1.two()}")

File: 2019/golang/1/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"strconv"
	"strings"
)

var (
	masses []Mass
)

type Mass int

func (m Mass) calcFuel() int {
	return Max(0, int(math.Floor(float64(m)/3))-2)
}

func Max(x, y int) int {
	if x < y {
		return y
	}
	return x
}

func main() {
	one()
	two()
}

func one() {
	prepareInput("input.txt")
	totalFuel := 0
	for _, mass := range masses {
		totalFuel += mass.calcFuel()
	}
	fmt.Printf("One: %d\n", totalFuel)
}

func two() {
	prepareInput("input.txt")
	totalFuel := 0
	for {
		newMasses := []Mass{}
		fuelCount := 0
		for _, mass := range masses {
			fuel := mass.calcFuel()
			fuelCount += fuel
			newMasses = append(newMasses, Mass(fuel))
		}
		totalFuel += fuelCount
		if fuelCount == 0 {
			break
		}
		masses = newMasses
	}
	fmt.Printf("Two: %d\n", totalFuel)
}

func prepareInput(file string) {
	masses = []Mass{}
	input, _ := ioutil.ReadFile(file)
	inputStrings := strings.Split(strings.TrimSpace(string(input)), "\r\n")
	for _, inputLine := range inputStrings {
		mass, _ := strconv.Atoi(inputLine)
		masses = append(masses, Mass(mass))
	}
}
